n-api: Re-use napi_env between modules

gabrielschulhof opened the pull request:
Create a std::map keyed on the isolate pointer and storing one napi_env
per isolate and protect it with a mutex.

Re: https://github.com/nodejs/node/pull/13872#discussion_r126908717

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
n-api

@bnoordhuis @mhdawson this is what sharing the `napi_env` would look like with a global static and a mutex.

bnoordhuis commented on src/node_api.cc:
```diff
@@ -705,6 +706,27 @@ bool FindWrapper(v8::Local<v8::Object> obj,
 
+std::map<v8::Isolate*, napi_env> napi_env_map;
+uv_once_t napi_env_once = UV_ONCE_INIT;
+uv_mutex_t napi_env_mutex;
```
`node::Mutex` (and use `node::Mutex::ScopedLock` below), then you won't need a uv_once_t either.

bnoordhuis commented on src/node_api.cc:
```diff
@@ -705,6 +706,27 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+  uv_mutex_lock(&napi_env_mutex);
+  result = napi_env_map[isolate];
+  if (result == nullptr) {
+    result = napi_env_map[isolate] = new napi_env__(isolate);
```
Are napi_env instances immortal?  If not, there should be code somewhere to remove it from the map again.

gabrielschulhof replied:
I guess we need a refcount then, because we need to know, for each isolate, how many modules are using the `napi_env`.

gabrielschulhof replied:
We shall also need a `v8impl::release_env()` to dereference/delete. However, these only makes sense when we have a module unload strategy.

gabrielschulhof replied:
I have changed the comment about `v8impl::get_env()` to remind of this.

bnoordhuis commented on test/addons-napi/test_env_sharing/compare_env.c:
```diff
@@ -0,0 +1,23 @@
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &external, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_external(env, external, &external_data));
+  NAPI_CALL(env,
+    napi_get_boolean(env, ((napi_env)external_data) == env, &return_value));
```
Style nit: can you line up the arguments in this file and others?

mhdawson commented on src/node_api.cc:
```diff
@@ -718,7 +740,7 @@ void napi_module_register_cb(v8::Local<v8::Object> exports,
   // Create a new napi_env for this module. Once module unloading is supported
   // we shall have to call delete on this object from there.
-  napi_env env = new napi_env__(context->GetIsolate());
+  napi_env env = v8impl::get_env(context->GetIsolate());
```
The comment should probably change now as we'd no longer delete the env when a module is unloaded.

mhdawson commented:
Can you change the title for the commit comment so that Re-use is lower case -> re-use

jasongin commented:
I think this isn't going to work when modules static-link `node_api.cc` from `node-addon-api` to target node versions without N-API built-in. In that case each module would have a separate `napi_env_map` instance.

gabrielschulhof commented:
@jasongin perhaps then we need a public API for per-isolate data that is better than what V8 offers.

gabrielschulhof commented:
@bnoordhuis @mhdawson I have now addressed your review comments.

I'm not sure I should be implementing the `napi_env` refcounting until we actually have a solution for unloading modules.

As for the interoperability with statically linked node-addon-api versions mentioned by @jasongin, I think we've already broken that with the stricter wrapping, because we perform the type check based on the value of the string pointer, not its contents.

jasongin commented:
> I think we've already broken that with the stricter wrapping, because we perform the type check based on the value of the string pointer, not its contents.

I think that is not actually a problem, because there should never be a need for a module to _unwrap_ an instance that was _wrapped_ by a different module.

gabrielschulhof commented:
@TimothyGu's [idea](https://github.com/nodejs/node/issues/14367) about using `v8::Private` to attach private properties to objects gave me the idea that we can place the `napi_env` in an `External` and attach it to the global object using such a `v8::Private` hidden property. That would obviate the need for a solution such as the one in this PR, and it would solve the problem of modules loaded via various incarnations of node-addon-api.

TBH, if we could attach such hidden properties to the global object we'd essentially have the per-~~isolate~~context storage mechanism we need.

In terms of implementation, from https://github.com/nodejs/nan/blob/master/nan_private.h it looks like `v8::Private` is available in node 8 and node 6 but not node 4. So, for node-addon-api we'd have to `#ifdef` the [way](https://github.com/nodejs/nan/blob/master/nan_private.h#L15-L22) nan `#ifdef`s to cover both cases. This means that the contents of node_api.cc would irrevocably diverge between node and node-addon-api.

RFC @bnoordhuis @mhdawson @jasongin

gabrielschulhof commented:
@jasongin I wouldn't dismiss the idea of cross-module wrapping/unwrapping so easily. I can imagine a scenario where we have two different packages depending on the same N-API module and the application passing objects originating from one to the other and vice versa.

gabrielschulhof commented:
@mhdawson @bnoordhuis @jasongin I re-wrote this patch using @TimothyGu's idea regarding `v8::Private`. This seems to be a much better solution. It's cleaner, it does not use a mutex, and it'll work across both built-in and external versions of N-API, provided the external version is ABI-compatible with the internal version - that is, we keep `node_api.cc` well in sync in order to make sure that the contents of an instance of `napi_env` is the same both externally and internally.

gabrielschulhof commented:
Oh, and it takes care of the eventuality of module unloading, and the eventuality of a module having to be loaded multiple times for multiple contexts as [foretold](https://github.com/nodejs/node/blob/master/src/node.cc#L2493-L2495) by @bnoordhuis.

gabrielschulhof commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+  auto maybe_key = v8::String::NewFromUtf8(isolate, "N-API Environment",
+      v8::NewStringType::kInternalized);
+  CHECK(!maybe_key.IsEmpty());
+  auto key = v8::Private::ForApi(isolate, maybe_key.ToLocalChecked());
```
That's a bit of a chicken-and-egg problem. I mean, I could create a `std::map` keyed on the isolate and store these `v8::Private`s but then I might as well go with the previous approach of storing the `napi_env` in the map's value.

This code runs once as part of `DLopen()` for a given module and then never again for that module. Given this, is it worth caching this value?

gabrielschulhof replied:
Actually, storing the `napi_env` instead of the v8::Private means I lose the ability to pass a `napi_env` created internally to node-addon-api, so I wouldn't revert to the previous approach. Still, I would only introduce a mutex if caching it really is worth it for the load-module use case.

TimothyGu replied:
If module loading performance isn't a concern, I wouldn't worry about it. FWIW it's also not necessary to use a mutex as `ForApi` is idempotent.

gabrielschulhof replied:
I thought the point of caching the `v8::Private` was to avoid calling `ForApi` more than once per new isolate. Or did you mean that I should cache the result of `v8::String::NewFromOneByte`?

If I use a `std::map` I must protect it with a mutex because another thread might be in the process of depositing a new value or it might be in the process of erasing an unused value.

I suppose you mean the kind of caching that I see in node.cc, like `env->exports_string()`. I guess maybe I should start looking into using `node::Environment`.

gabrielschulhof commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+    auto external = v8::External::New(isolate, result);
+    auto maybe_set = global->SetPrivate(context, key, external);
+    CHECK(maybe_set.IsJust());
+    CHECK(maybe_set.FromJust());
```
AFAICT `FromJust()` doesn't take any parameters. I tried to combine, but I get
```
In file included from ../src/node_internals.h:28:0,
                 from ../src/node.h:173,
                 from ../src/node_buffer.h:25,
                 from ../src/node_api.cc:11:
../src/node_api.cc: In function ‘napi_env__* {anonymous}::v8impl::get_env(v8::Local<v8::Context>)’:
../src/node_api.cc:735:35: error: no matching function for call to ‘v8::Maybe<bool>::FromJust(bool)’
     CHECK(maybe_set.FromJust(false));
                                   ^
../src/util.h:107:44: note: in definition of macro ‘UNLIKELY’
 #define UNLIKELY(expr) __builtin_expect(!!(expr), 0)
                                            ^~~~
../src/node_api.cc:735:5: note: in expansion of macro ‘CHECK’
     CHECK(maybe_set.FromJust(false));
     ^~~~~
In file included from ../src/node.h:63:0,
                 from ../src/node_buffer.h:25,
                 from ../src/node_api.cc:11:
../deps/v8/include/v8.h:8014:15: note: candidate: T v8::Maybe<T>::FromJust() const [with T = bool]
   V8_INLINE T FromJust() const {
               ^~~~~~~~
../deps/v8/include/v8.h:8014:15: note:   candidate expects 0 arguments, 1 provided
```

TimothyGu replied:
Oops I meant `FromMaybe(false)`

TimothyGu commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+  delete data.GetParameter();
+}
+
+napi_env get_env(v8::Local<v8::Context> context) {
```
All other functions in the v8impl namespace uses CamelCase. Might be good to do that here etc.

TimothyGu commented on test/addons-napi/test_env_sharing/test.js:
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+const common = require('../../common');
+const store_env = require(`./build/${common.buildType}/store_env`);
```
camelCase in JS

bnoordhuis commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+  auto global = context->Global();
+
+  auto maybe_key = v8::String::NewFromOneByte(isolate,
+      reinterpret_cast<const uint8_t *>("N-API Environment"),
```
Nit: no space before `*`

bnoordhuis commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+
+  auto maybe_value = global->GetPrivate(context, key);
+  CHECK(!maybe_value.IsEmpty());
+  auto value = maybe_value.ToLocalChecked();
```
Just an observation but .ToLocalChecked() also does the equivalent of `CHECK(!maybe_value.IsEmpty())`.

bnoordhuis commented on src/node_api.cc:
```diff
@@ -706,6 +706,40 @@ bool FindWrapper(v8::Local<v8::Object> obj,
+    auto external = v8::External::New(isolate, result);
+    auto maybe_set = global->SetPrivate(context, key, external);
+    CHECK(maybe_set.FromMaybe(false));
+    v8::Persistent<v8::External> persistent(isolate, external);
```
This works because `~Persistent()` currently doesn't call .Reset() but v8.h indicates that may change in the future.  Not sure what to do about that but I figured you should know.

bnoordhuis commented on test/addons-napi/test_env_sharing/compare_env.c:
```diff
@@ -0,0 +1,22 @@
+
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &external, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_external(env, external, &data));
+  NAPI_CALL(env, napi_get_boolean(env, ((napi_env)data) == env, &return_value));
```
static_cast?

gabrielschulhof replied:
This is C.

gabrielschulhof commented:
@TimothyGu I have updated to camelCase.

@bnoordhuis I have removed the `Persistent` and replaced it with one of our self-destructing `v8impl::Reference` objects.

gabrielschulhof commented:
I change the commit message subject to contain no uppercase letters.

gabrielschulhof commented:
If this is good to go, may I have a CI run please?

TimothyGu commented:
Seems like the CI is having some issues, though.

gabrielschulhof commented:
I think the CI is fixed now. Could somebody please submit this PR?

gabrielschulhof commented:
Oh, NM. Thanks!

gabrielschulhof commented:
https://ci.nodejs.org/job/node-test-pull-request/9351/ completed successfully a while ago. It shouldn't still show as pending in this PR.

addaleax commented:
@gabrielschulhof CI status reporting is broken, don’t worry about it (not sure whether somebody’s looking into it or not).

gabrielschulhof commented:
@addaleax NP. I'm just eager to get this landed, that's all :)

addaleax commented:
I’m going to sleep soon but if it’s not landed tomorrow I can do that.

Also, just nominated you for push & CI access, that should be a bit more practical if you like :)

gabrielschulhof commented:
Wow! I'm honoured! Thanks!

TimothyGu commented:
Landed in 9926dfebc88c22a1a9bdef55f7530c74124f63aa.

I also amended the commit message to reference https://github.com/nodejs/node/issues/14367 rather than https://github.com/nodejs/node/pull/13872#discussion_r126908717, as the approach we ended up taking was documented in #14367.

TimothyGu commented:
And also, I'd second Anna's nomination :)
